src/Type/Php/GetDefinedVarsFunctionReturnTypeExtension.php

```diff
@@ -0,0 +1,68 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use function array_filter;
+use function array_map;
+use function array_merge;
+use function array_values;
+use function count;
+use function range;
+
+// based on code by @ruudk
+final class GetDefinedVarsFunctionReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'get_defined_vars';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if ($scope->canAnyVariableExist()) {
+			return new ArrayType(
+				new StringType(),
+				new MixedType(),
+			);
+		}
+
+		$variables = array_values(array_filter(
+			$scope->getDefinedVariables(),
+			static fn ($variable) => $variable !== 'this',
```

Comment:
Not sure if we need to remove `$this`. 

I used that in my example snippet, because I didn't want `this` in there. But in general cases, I think you do want to see it when it's there.

Reply:
ok I removed it